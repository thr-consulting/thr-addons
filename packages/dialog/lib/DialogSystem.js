// @flow

/* eslint-disable react/no-did-update-set-state */
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import type {ChildrenArray, Element} from 'react';
import Dock from 'react-dock';
import styles from './dialog.css';

const defaultOptions = {
	position: 'bottom',
	isVisible: false,
	dimMode: 'opaque',
	zIndex: 9997,
};

type Props = {
	children: ChildrenArray<*>,
};

type State = {
	position: string,
	isVisible: boolean,
	dimMode: string,
	zIndex: number,
	component: ?Element<*>,
	height: number,
};

/**
 * Wrap your app with a DialogSystem instance to make dialogs available. You only need one of these per app.
 * @class
 * @property {Component[]} children - Your application components.
 */
export default class DialogSystem extends Component<Props, State> {
	static childContextTypes = {
		showDialog: PropTypes.func,
		closeDialog: PropTypes.func,
	};

	constructor(props: Object) {
		super(props);
		this.state = {
			...defaultOptions,
			component: null,
			height: 0,
		};
	}

	getChildContext() {
		return {
			showDialog: this.showDialog,
			closeDialog: this.closeDialog,
		};
	}

	componentDidUpdate() {
		if (this._contentNode) {
			const height = this._contentNode.offsetHeight;

			if (height !== this.state.height) {
				this.setState({
					height,
				});
			}
		}
	}

	props: Props;
	_contentNode: ?HTMLElement;

	/**
	 * Shows a dialog. This method is available via the React Context.
	 * @function showDialog
	 * @memberOf DialogSystem
	 * @tag Context
	 * @param {Element} component - The Dialog component to display.
	 * @param {object} options - Custom React Dock props.
	 */
	showDialog = (component: Element<*>, options: Object = defaultOptions) => {
		this.setState({
			...options,
			component,
			isVisible: true,
			height: 0,
		});
	};

	closeDialog = () => {
		this.setState({
			isVisible: false,
			component: null, // TODO This hides (removes) the dialog component before it slides down.
		});
	};

	render() {
		return (
			<div className={styles.dialogSystem}>
				{this.props.children}
				<Dock
					position={this.state.position}
					isVisible={this.state.isVisible}
					dimMode={this.state.dimMode}
					zIndex={this.state.zIndex}
					size={this.state.height}
					fluid={false}
				>
					<div ref={r => (this._contentNode = r)}>
						{this.state.component ? React.cloneElement(this.state.component, {
							showDialog: this.showDialog,
							closeDialog: this.closeDialog,
						}) : null}
					</div>
				</Dock>
			</div>
		);
	}
}
