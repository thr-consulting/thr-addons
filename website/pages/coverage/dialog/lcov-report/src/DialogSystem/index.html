<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/DialogSystem</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="/lcov/prettify.css" />
    <link rel="stylesheet" href="/lcov/base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(/lcov/sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/DialogSystem
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.33% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>10/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">71.43% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>5/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">85.71% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>6/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.33% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>10/12</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="DialogSystem.js"><a href="DialogSystem.js.html">DialogSystem.js</a></td>
	<td data-value="83.33" class="pic high"><div class="chart"><div class="cover-fill" style="width: 83%;"></div><div class="cover-empty" style="width:17%;"></div></div></td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Oct 04 2018 09:33:27 GMT-0500 (CDT)
</div>
</div>
<script src="/lcov/prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="/lcov/sorter.js"></script>
<script src="/lcov/block-navigation.js"></script>
</body>
</html>
