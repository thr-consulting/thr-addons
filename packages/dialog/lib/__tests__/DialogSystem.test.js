import React from 'react';
import PropTypes from 'prop-types';
import {shallow, mount} from 'enzyme';
import DialogSystem from '../DialogSystem';

function HelloWorld(props) {
	props.func();
	return (
		<div>Hello World</div>
	);
}
HelloWorld.propTypes = {
	func: PropTypes.func,
};

class DialogSystemFunctional extends React.Component {
	static contextTypes = {
		showDialog: PropTypes.func,
	};

	static propTypes = {
		func: PropTypes.func,
	};

	handleClick = () => {
		this.context.showDialog(<HelloWorld func={this.props.func}/>);
	}

	render() {
		return <button onClick={this.handleClick}>Click Me</button>;
	}
}

describe('DialogSystem', () => {
	it('should render without throwing an error', () => {
		expect(shallow(
			<DialogSystem/>
		)).toMatchSnapshot();
	});

	it('should respond to height changes', () => {
		const func = jest.fn();
		const wrapper = mount(
			<DialogSystem>
				<DialogSystemFunctional func={func}/>
			</DialogSystem>
		);
		const button = wrapper.find('button');
		button.simulate('click');
		expect(func.mock.calls.length).toBe(1);
	});

	// it('should respond to changes', () => {
	// 	const mockOnChange = jest.fn();
	//
	// 	const wrapper = mount(
	// 		<FieldMap
	// 			value="initial value"
	// 			as={Input}
	// 			asValue="value"
	// 			onChange={mockOnChange}
	// 		/>);
	// 	const u = wrapper.find('input');
	// 	u.simulate('change', {target: {value: 'blah blah'}});
	//
	// 	expect(mockOnChange.mock.calls.length).toBe(1);
	// 	expect(mockOnChange.mock.calls[0][0]).toEqual('blah blah');
	// });
});
