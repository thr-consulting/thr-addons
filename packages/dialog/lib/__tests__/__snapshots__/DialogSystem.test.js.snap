// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogSystem should render without throwing an error 1`] = `
<div
  className="dialogSystem"
>
  <Dock
    defaultSize={0.3}
    dimMode="opaque"
    duration={200}
    fluid={false}
    isVisible={false}
    position="bottom"
    size={0}
    zIndex={9997}
  >
    <div />
  </Dock>
</div>
`;
